src/appleseed/renderer/modeling/aov/diagnosticaov.cpp

```diff
@@ -0,0 +1,501 @@
+            const size_t height = tile.get_height();
+
+            for (size_t y = 0; y < height; ++y)
+            {
+                for (size_t x = 0; x < width; ++x)
+                {
+                    Color<uint8, 1> sample_state;
+                    m_invalid_sample_tile->get_pixel(x, y, sample_state);
+
+                    Color3f color;
+                    Color4f beauty_color;
+
+                    switch (sample_state[0])
+                    {
+                      case NoState:
+                        color = Color3f(1.0f, 0.0f, 0.0f);
+                        break;
+
+                      case InvalidSample:
+                        color = Color3f(1.0f, 0.0f, 1.0f);
+                        break;
+
+                      case CorrectSample:
+                        tile.get_pixel(x, y, beauty_color);
+                        color.set(0.2f * luminance(beauty_color.rgb()));     // 20% of luminance
+                        break;
+
+                      assert_otherwise;
+                    }
+
+                    aov_tile.set_pixel(x, y, color);
+                }
+            }
+        }
+
+        void on_pixel_begin(const Vector2i& pi) override
+        {
+            m_invalid_sample_count = 0;
+        }
+
+        void on_pixel_end(const Vector2i& pi) override
+        {
+            // Store a hint corresping to the sample state in the tile.
+            if (pi.x >= m_tile_origin_x &&
+                pi.y >= m_tile_origin_y &&
+                pi.x <= m_tile_end_x &&
+                pi.y <= m_tile_end_y)
+            {
+                const Vector2i pt(pi.x - m_tile_origin_x, pi.y - m_tile_origin_y);
+                m_invalid_sample_tile->set_pixel(pt.x, pt.y,
+                    m_invalid_sample_count > 0 ? &InvalidSample : &CorrectSample);
+            }
+        }
+
+        void write(
+            const PixelContext&         pixel_context,
+            const ShadingPoint&         shading_point,
+            const ShadingComponents&    shading_components,
+            const AOVComponents&        aov_components,
+            ShadingResult&              shading_result) override
+        {
+            // Detect invalid samples.
+            if (!shading_result.is_valid())
+                m_invalid_sample_count++;
+        }
+
+      private:
+        Image&                                  m_image;
+        size_t                                  m_invalid_sample_count;
+        std::unique_ptr<foundation::Tile>       m_invalid_sample_tile;
+        int                                     m_tile_origin_x;
+        int                                     m_tile_origin_y;
+        int                                     m_tile_end_x;
+        int                                     m_tile_end_y;
+    };
+
+    //
+    // Diagnostic AOV.
+    //
+
+    class DiagnosticAOV
+      : public AOV
+    {
+      public:
+        explicit DiagnosticAOV(const char* name, const ParamArray& params)
```

Comment:
Remove `explicit`.

Reply:
Why is this AOV named while the others aren't?

Reply:
Something I forgot to remove.

Reply:
Because other AOVs inherit from this one.

Reply:
Why removing explicit ?

Reply:
Because it's not useful for constructors with more than one argument... Or at least that was true until C++11: https://stackoverflow.com/a/39122237/393756

It seems that it makes sense starting with C++11 to mark all constructors as explicit... Today I learn!

You can leave `explicit` then. If you already remove it then no big deal, it's fine too.